File: code/ex39.java
class Solution {
    public int n;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        n = candidates.length;
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(res, path, candidates, target, 0);
        return res;
    }
    
    private void dfs(List<List<Integer>> res, List<Integer> path, int[] candidates, int target, int s) {
        if (target < 0)    return;
        if (target == 0) {
            res.add(new ArrayList<Integer>(path));
        }
        
        for (int i = s; i < n; i++) {
            if (candidates[i] > target)    continue;
            path.add(candidates[i]);
            dfs(res, path, candidates, target - candidates[i], i);
            path.remove(path.size() - 1);
        }
    }
}

File: code/ex39.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> cur;
        dfs(candidates, res, cur, target, 0);
        return res;
    }
    
private:
    void dfs(vector<int>& candidates, vector<vector<int>>& res, vector<int>& cur, int target, int s) {
        if (target <= 0) {
            if (target == 0)    res.push_back(cur);
            return;
        }
        
        for (int i = s; i < candidates.size(); i++) {
            if (i > s && candidates[i] == candidates[i - 1])    continue;
            cur.push_back(candidates[i]);
            dfs(candidates, res, cur, target - candidates[i], i);
            cur.pop_back();
        }
    }
};